politeiawww/prices.go

```diff
@@ -0,0 +1,122 @@
+package main
+
+import (
+	"encoding/json"
+	"net/http"
+	"strconv"
+	"time"
+
+	cms "github.com/decred/politeia/politeiawww/api/cms/v1"
+)
+
+const poloURL = "https://poloniex.com/public"
+const httpTimeout = time.Second * 3
+const pricePeriod = 900
+
+type poloChartData struct {
+	Date            uint64  `json:"date"`
+	WeightedAverage float64 `json:"weightedAverage"`
+}
+
+// GetMonthAverage returns the average USD/DCR price for a given month
+func (p *politeiawww) GetMonthAverage(month time.Month, year int) (float64, error) {
+	startTime := time.Date(year, month, 1, 0, 0, 0, 0, time.UTC)
+	endTime := startTime.AddDate(0, 1, 0)
+
+	unixStart := startTime.Unix()
+	unixEnd := endTime.Unix()
+
+	// Download BTC/DCR and USDT/BTC prices from Poloniex
+	dcrPrices, err := getPrices("BTC_DCR", unixStart, unixEnd)
+	if err != nil {
+		return 0, err
+	}
+	btcPrices, err := getPrices("USDT_BTC", unixStart, unixEnd)
+	if err != nil {
+		return 0, err
+	}
+
+	// Create a map of unix timestamps => average price
+	usdtDcrPrices := make(map[uint64]float64)
+
+	// Select only timestamps which appear in both charts to
+	// populate the result set. Multiply BTC/DCR rate by
+	// USDT/BTC rate to get USDT/DCR rate.
+	for timestamp, dcr := range dcrPrices {
+		if btc, ok := btcPrices[timestamp]; ok {
+			usdtDcrPrices[timestamp] = dcr * btc
+		}
+	}
+
+	// Calculate and return the average of all USDT/DCR prices
+	var average float64
+	for _, price := range usdtDcrPrices {
+		average += price
+	}
+	average = average / float64(len(usdtDcrPrices))
+
+	// Polo doesn't like >6 requests per second
+	time.Sleep(300 * time.Millisecond)
```

Comment:
I just want to highlight this line to make sure you are aware of it. I put this here as a cheap hack, and with the way my app is configured it works just fine.

I'm not aware of exactly how this code is being utilized within CMS, so it could be a non-issue, but just wanted to raise awareness.